nova/compute/manager.py

```diff
@@ -10680,6 +10681,35 @@ def _reclaim_queued_deletes(self, context):
                                 "instance: %s",
                                 e, instance=instance)
 
+    @periodic_task.periodic_task(
+            spacing=CONF.instances_stuck_in_deleting_cleanup_interval)
+    def _cleanup_instances_stuck_in_deleting(self, context):
+        """Reclaim instances stuck in DELETING state, but not really destroyed,
+        and retrigger terminate_instance for them.
+        """
+        filters = {'task_state': task_states.DELETING,
+                   'host': self.host}
+        instances = objects.InstanceList.get_by_filters(
+            context, filters,
+            expected_attrs=objects.instance.INSTANCE_DEFAULT_FIELDS,
+            use_slave=True)
+        for instance in instances:
+            # Check if the instance has a process lock
+            for name, s in _semaphores._semaphores.items():
+                if name == instance.uuid:
+                    break
+            else:
```

Comment:
What if your list of instances is outdated after you checked the first couple of locks? We then call `terminate_instance()` with an already deleted `Instance` object. Can the code handle that?

Reply:
yes it seems that the [`terminate_instance()` ](https://github.com/sapcc/nova/blob/db78134420a4aa30c815d3fd3b465fd8aeb4d840/nova/compute/manager.py#L3349)can handle that .. it will report that the instance disappeared but it will not fail